Subclass Kubevirt InfraManager to separate Openshift Virtualization provider

nasark opened the pull request:
We can subclass `Kubevirt::InfraManager` to create a separate provider for Openshift Virtualization

Depends on:
- [ ] https://github.com/ManageIQ/manageiq/pull/23165
- [ ] https://github.com/ManageIQ/manageiq-providers-kubevirt/pull/247
- [ ] https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/537

@miq-bot assign @agrare 
@miq-bot add_label enhancement, radjabov/yes?
@miq-bot add_reviewer @agrare

nasark commented:
@agrare WIP - need to test with an Openshift Virtualization provider and clean up specs. Feel free to review what is here though

Fryguy commented on app/models/manageiq/providers/openshift/container_manager.rb:
```diff
@@ -3,6 +3,13 @@
 class ManageIQ::Providers::Openshift::ContainerManager < ManageIQ::Providers::Kubernetes::ContainerManager
   DEFAULT_EXTERNAL_LOGGING_ROUTE_NAME = "logging-kibana-ops".freeze
 
+  has_one :infra_manager,
```
Does it makes sense for this to just live in the base class, or is this intentionally overriding the class name?

agrare replied:
This is here because we have to override the class_name to be Openshift::InfraManager so the correct class gets created

Fryguy commented on app/models/manageiq/providers/openshift/infra_manager.rb:
```diff
@@ -0,0 +1,74 @@
+  supports :provisioning
+
+  def self.ems_type
+    @ems_type ||= "openshift".freeze
```
I think this has to be different than the container manager, so maybe something like `openshift_virt` or `openshift_infra`?

agrare replied:
:+1: openshift_infra would be good (OpenStack undercloud/InfraManager is openstack_infra)

Fryguy commented on app/models/manageiq/providers/openshift/infra_manager/refresh_worker/runner.rb:
```diff
@@ -0,0 +1,168 @@
+  # @return [ManageIQ::Providers::Kubevirt::InfraManager] The manager.
+  #
+  def manager
+    @manager ||= ManageIQ::Providers::Openshift::InfraManager.find(@cfg[:ems_id])
```
I'm surprised so much of this overriding is necessary.  Makes me wonder if there's a better way, such as providing a method like `manageiq_class`, which returns `ManageIQ::Providers::Openshift::InfraManager`, and then the actual logic would be something like

```
    @manager ||= manager_class.find(@cfg[:ems_id])
```

Fryguy replied:
This is the simple example of course - the same comment really applies to all of the methods in this file.

Fryguy commented:
Very cool!  I like that we didn't need a new repository, and it's just included in the openshift repo.

Fryguy commented:
One other thing I think we need is the form for adding a provider needs to change so you can only select OpenShift Virtualization under the Virtualization tab.

agrare commented on app/models/manageiq/providers/openshift/infra_manager.rb:
```diff
@@ -0,0 +1,74 @@
+class ManageIQ::Providers::OpenShift::InfraManager < ManageIQ::Providers::Kubevirt::InfraManager
+  belongs_to :parent_manager,
+             :foreign_key => :parent_ems_id,
+             :class_name  => "ManageIQ::Providers::ContainerManager"
```
This should be `ManageIQ::Providers::Openshift::ContainerManager`, then you can add an invere_of :infra_manager

agrare commented on app/models/manageiq/providers/openshift/infra_manager.rb:
```diff
@@ -0,0 +1,74 @@
+             :foreign_key => :parent_ems_id,
+             :class_name  => "ManageIQ::Providers::ContainerManager"
+
+  delegate :authentication_check,
+           :authentication_for_summary,
+           :authentication_token,
+           :authentications,
+           :endpoints,
+           :default_endpoint,
+           :zone,
+           :to        => :parent_manager,
+           :allow_nil => true
```
I think all of these can be inherited from the base class

agrare commented on app/models/manageiq/providers/openshift/infra_manager.rb:
```diff
@@ -0,0 +1,74 @@
+  end
+
+  def authentication_for_providers
+    authentications.where(:authtype => :openshift)
```
Not necessarily for this PR but if most of these overrides are for `:authtype => :openshift` we could use e.g. a DEFAULT_AUTH_TYPE constant and switch to that in the base Kubevirt class

agrare replied:
e.g.
```
  def connect(opts = {}) 
    # Get the authentication token:
    token = opts[:token] || authentication_token(:kubevirt)
```
If we changed that to `authentication_token(default_authentication_type)`

agrare commented on app/models/manageiq/providers/openshift/infra_manager.rb:
```diff
@@ -0,0 +1,74 @@
+  #
+  def connect(opts = {})
+    # Get the authentication token:
+    token = opts[:token] || authentication_token(:openshift)
```
This is a good example, this whole method is probably copy&paste except for this `:openshift` here

agrare commented on app/models/manageiq/providers/openshift/infra_manager.rb:
```diff
@@ -0,0 +1,74 @@
+    )
+  end
+
+  def virtualization_endpoint
+    connection_configurations.kubevirt.try(:endpoint)
+  end
```
This looks identical to the base class

agrare commented on app/models/manageiq/providers/openshift/infra_manager/template.rb:
```diff
@@ -0,0 +1,15 @@
+ManageIQ::Providers::Kubevirt::InfraManager::Template.include(ActsAsStiLeafClass)
+
+class ManageIQ::Providers::Openshift::InfraManager::Template < ManageIQ::Providers::Kubevirt::InfraManager::Template
+  supports :provisioning do
+    if ext_management_system
+      ext_management_system.unsupported_reason(:provisioning)
+    else
+      _('not connected to ems')
+    end
+  end
```
Is this different from the base class?

agrare commented:
Not sure if you want to do this in another PR but we should make the "virtualization" endpoint dropdown option (https://github.com/ManageIQ/manageiq-providers-kubernetes/blob/master/app/models/manageiq/providers/kubernetes/container_manager.rb#L331-L342) in the Openshift::ContainerManager subclass only show "Openshift Virtualization"

agrare commented:
I think we should get some VCR specs in here, kubevirt has some weird mocking that they do for their refresher specs I'm not sure I want to copy what they do but a "simple" refresher spec doing a full refresh should be sufficient.

nasark commented:
@agrare I think I've addressed the feedback, please have another look. Linked the relevant PRs

Still working on the specs though

agrare commented on app/models/manageiq/providers/openshift/container_manager.rb:
```diff
@@ -21,6 +28,20 @@ def self.event_monitor_class
     ManageIQ::Providers::Openshift::ContainerManager::EventCatcher
   end
 
+  def virtualization_options
```
```suggestion
  def self.virtualization_options
```

agrare commented on app/models/manageiq/providers/openshift/infra_manager/refresh_worker.rb:
```diff
@@ -0,0 +1,2 @@
+class ManageIQ::Providers::Openshift::InfraManager::RefreshWorker < ManageIQ::Providers::Kubevirt::InfraManager::RefreshWorker
```
We don't need to inherit anything from the Kubevirt RefreshWorker, and doing this nests the settings under kubevirt so we typically don't do this for simple workers (e.g. see Openshift::ContainerManager::RefreshWorker it doesn't < Kubernetes)
```suggestion
class ManageIQ::Providers::Openshift::InfraManager::RefreshWorker < ManageIQ::Providers::BaseManager::RefreshWorker
```

agrare replied:
You also need to add a settings section for this new worker to config/settings.yml
```
diff --git a/config/settings.yml b/config/settings.yml
index dc097791..448ed39b 100644
--- a/config/settings.yml
+++ b/config/settings.yml
@@ -33,3 +33,4 @@
           :ems_metrics_collector_worker_openshift: {}
         :ems_refresh_worker:
           :ems_refresh_worker_openshift: {}
+          :ems_refresh_worker_openshift_infra: {}
```

agrare replied:
NOTE we'll also need to add .service and .target files for the new openshift-infra-refresh worker in the systemd/ directory here (specs will catch this but just want to head it off now)

agrare commented:
When running the full worker I see the following error:
```
[----] I, [2024-08-27T14:07:56.049629#148737:8890]  INFO -- evm: MIQ(ManageIQ::Providers::Openshift::ContainerManager::EventCatcher::Runner#start_event_monitor) EMS [api.crc.testing] as [] Starting Event Monitor Thread
[----] I, [2024-08-27T14:07:56.050809#148737:8890]  INFO -- evm: MIQ(ManageIQ::Providers::Openshift::ContainerManager::EventCatcher::Runner#start_event_monitor) EMS [api.crc.testing] as [] Started Event Monitor Thread
[----] E, [2024-08-27T14:07:56.641440#148735:8890] ERROR -- evm: MIQ(ManageIQ::Providers::Openshift::ContainerManager::RefreshWorker::Runner#refresh_block) EMS [CRC], id: [6] Refresh failed: undefined method `vms' for #<ManageIQ::Providers::Openshift::Inventory::Collector::ContainerManager
[----] E, [2024-08-27T14:07:56.642961#148735:8890] ERROR -- evm: /home/grare/adam/src/manageiq/manageiq-providers-kubevirt/app/models/manageiq/providers/kubevirt/inventory/parser/full_refresh.rb:8:in `parse'
/home/grare/adam/src/manageiq/manageiq/app/models/manageiq/providers/inventory.rb:38:in `block in parse'
/home/grare/adam/src/manageiq/manageiq/app/models/manageiq/providers/inventory.rb:35:in `each'
/home/grare/adam/src/manageiq/manageiq/app/models/manageiq/providers/inventory.rb:35:in `parse'
```

When using `>> EmsRefresh.refresh(ManageIQ::Providers::Openshift::InfraManager.first)` I see the same `ManageIQ::Providers::Kubevirt::InfraManager::Host is not a subclass of ManageIQ::Providers::Openshift::InfraManager::Host (ManageIQ::Providers::BaseManager::Refresher::PartialRefreshError)` that you do.

nasark commented:
@agrare container manager refresh is fixed now, still figuring out the OSV refresh error though:
```
ManageIQ::Providers::Kubevirt::InfraManager::Host is not a subclass of ManageIQ::Providers::Openshift::InfraManager::Host 
```

nasark commented:
- fixed refresh bugs for container provider and OSV
- added systemd files for `openshift_infra_manager_refresh`
- added specs (no VCR yet due to factory error)
- added OSV logo

nasark commented on app/models/manageiq/providers/openshift/inventory/parser/infra_manager/full_refresh.rb:
```diff
@@ -0,0 +1,20 @@
+class ManageIQ::Providers::Openshift::Inventory::Parser::InfraManager::FullRefresh < ManageIQ::Providers::Kubevirt::Inventory::Parser::FullRefresh
+
+  private
+
+  def provider_class
+    ManageIQ::Providers::Openshift
+  end
+
+  def product
+    "OpenShift Virtualization"
+  end
+
+  def vendor
+    "openshift_infra"
+  end
+
+  def version
+    '0.1.0'
+  end
```
These will probably need to be added to the `::PartialRefresh` and `::PartialTargetRefresh` classes since they use the common kubevirt parser class https://github.com/ManageIQ/manageiq-providers-kubevirt/pull/247/files#diff-ad8b5bb422b15e9a6a52b7f4ded177f8bce041245f2a05d5dd983a09f6e055c6R68-R72

@agrare Should we move these to a common mixin and include to the classes that need it? Kind of like https://github.com/ManageIQ/manageiq-providers-kubevirt/blob/master/app/models/manageiq/providers/kubevirt/constants.rb, or open to any ideas you may have

agrare replied:
Since these are static values we could move them to the persister definition as a default value.

agrare replied:
For example ovirt does a similar trick to work with the RHV subclass: https://github.com/ManageIQ/manageiq-providers-ovirt/blob/master/app/models/manageiq/providers/ovirt/inventory/persister/infra_manager.rb#L9-L11

nasark commented on spec/factories/ext_management_system.rb:
```diff
@@ -5,4 +5,9 @@
     end
   end
+
+  factory :ems_openshift_infra,
+          :aliases => ["manageiq/providers/openshift/infra_manager"],
+          :class   => "ManageIQ::Providers::Openshift::InfraManager",
+          :parent => :ems_kubevirt
```
Hitting some issues with this factory hence no VCRs at the moment:
```
  1) ManageIQ::Providers::Openshift::Inventory::Parser::InfraManager::FullRefresh#refresh works correctly with one node
     Failure/Error:
       FactoryBot.create(:ems_openshift_infra,
                         :name => "openshift Virtualization Manager",
                         :connection_configurations => [{:endpoint       => {:role              => :default,
                                                                             :hostname          => host,
                                                                             :port              => port,
                                                                             :security_protocol => "ssl-without-validation"},
                                                         :authentication => {:role     => :bearer,
                                                                             :auth_key => token,
                                                                             :userid   => "_"}}])
     
     ActiveRecord::AssociationTypeMismatch:
       ManageIQ::Providers::Openshift::ContainerManager(#80160) expected, got #<ManageIQ::Providers::Kubernetes::ContainerManager
```

agrare commented on app/models/manageiq/providers/openshift/container_manager.rb:
```diff
@@ -21,9 +28,23 @@ def self.event_monitor_class
   def self.raw_connect(hostname, port, options)
     options[:service] ||= "openshift"
-    send("#{options[:service]}_connect", hostname, port, options)
+    send(:"#{options[:service]}_connect", hostname, port, options)
```
Is this necessary?

nasark replied:
Seems like my rubocop did this automatically haha

agrare commented on spec/vcr_cassettes/full_refresh.yml:
Usually we like to namespace these under the provider name e.g. https://github.com/ManageIQ/manageiq-providers-openshift/tree/master/spec/vcr_cassettes/manageiq/providers/openshift/container_manager

agrare commented on spec/models/manageiq/providers/openshift/infra_manager/full_refresh_spec.rb:
Prefer this matching the Refresher class name

agrare commented on spec/models/manageiq/providers/openshift/infra_manager/full_refresh_spec.rb:
```diff
@@ -0,0 +1,91 @@
+require 'json'
+require 'recursive_open_struct'
+
+describe ManageIQ::Providers::Openshift::Inventory::Parser::InfraManager::FullRefresh do
```
```suggestion
describe ManageIQ::Providers::Openshift::InfraManager::Refresher do
```

agrare commented on spec/vcr_cassettes/full_refresh.yml:
Hm seem to have lost my comment, but typically we want this to match the provider class namespace like `spec/vcr_cassettes/manageiq/providers/openshift/container_manager/refresher.yml`
You can use `VCR.use_cassette(described_class.name.underscore)`

agrare commented:
@miq-bot cross-repo-tests https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/537 https://github.com/ManageIQ/manageiq-providers-kubevirt/pull/247 https://github.com/ManageIQ/manageiq/pull/23165

Fryguy commented on app/assets/images/svg/vendor-openshift_infra.svg:
Should this live in manageiq-decorators?  I prefer it in the plugin class, but wasn't sure if it was required over there.

Fryguy replied:
On that note, in manageiq-decorators we use symlinks when there are multiple images that point to the same thing, so this could just be a symlink over there.

nasark replied:
It doesn't seem like it's required in manageiq-decorators

<img width="644" alt="Screenshot 2024-08-30 at 12 22 38 PM" src="https://github.com/user-attachments/assets/cc5cd885-2b75-48a9-862a-0ec53b86e184">

nasark commented:
@miq-bot cross-repo-tests https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/537 https://github.com/ManageIQ/manageiq-providers-kubevirt/pull/247 https://github.com/ManageIQ/manageiq/pull/23165

nasark commented:
@agrare @Fryguy cross repo tests are green https://github.com/ManageIQ/manageiq-cross_repo-tests/pull/911

agrare commented on app/models/manageiq/providers/openshift/inventory.rb:
```diff
@@ -1,2 +1,39 @@
+    if !ems.kind_of?(EmsInfra)
+      super
+    else
+      parser_type = if target_is_vm?(target)
+                      "PartialTargetRefresh"
+                    else
+                      "FullRefresh"
+                    end
```
Really minor for a follow-up but something this simple taking up 5 lines could be a ternary operator instead on one line
```suggestion
      parser_type = target_is_vm?(target) ? "PartialTargetRefresh" : "FullRefresh"
```

I'm not usually a "hey you can turn this big multi-step thing into one looong line" guy but this is pretty straightforward

agrare commented on spec/models/manageiq/providers/openshift/infra_manager/refresher_spec.rb:
```diff
@@ -0,0 +1,92 @@
+      end
+    end
+
+    it 'works correctly with one node' do
```
On that note, a targeted refresh spec would be a good addition in the future

Fryguy commented:
Backported to `radjabov` in commit 0ccbcc988c82a8569943b8ddad947f0a8e595df1.

```text
commit 0ccbcc988c82a8569943b8ddad947f0a8e595df1
Author: Adam Grare <adam@grare.com>
Date:   Tue Sep 10 11:37:39 2024 -0400

    Merge pull request #264 from nasark/subclass_kubevirt
    
    Subclass Kubevirt InfraManager to separate Openshift Virtualization provider
    
    (cherry picked from commit 2d66f57e7b3838c20888243a53e8579f7606aa5b)
```
